lib/wallaroo/core/source/connector_source/connector_framed_source_notify.pony

```diff
@@ -82,91 +254,348 @@ class ConnectorSourceNotify[In: Any val]
+      | let m: cwm.MessageMsg =>
+        // check that we're in state that allows processing messages
+        if _fsm_state isnt _ProtoFsmStreaming then
+          return _to_error_state(source, "Bad protocol FSM state")
+        end
+        if not cwm.FlagsAllowed(m.flags) then
+          return _to_error_state(source, "Bad MessageMsg flags")
+        end
 
-          if decoded isnt None then
-            _runner.run[In](_pipeline_name, pipeline_time_spent, decoded,
-              consume initial_key, ingest_ts, _watermark_ts, _source_id,
-              source, _router, msg_uid, None, decode_end_ts,
-              latest_metrics_id, ingest_ts, _metrics_reporter)
-          else
-            (true, ingest_ts)
-          end
+        // try to process message
+        if not _active_streams.contains(m.stream_id) then
+          return _to_error_state(source, "Bad stream_id " + m.stream_id.string())
         else
-          @printf[I32](("Unable to decode message at " + _pipeline_name +
-            " source\n").cstring())
-          ifdef debug then
-            Fail()
+          try
+            let s = _active_streams(m.stream_id)?
+            let msg_id = try
+              m.message_id as cwm.MessageId
+            else
+              0
+            end
+
+            if (msg_id > 0) and (msg_id <= s.last_seen) then
+              // skip processing of an already seen message
+              return _continue_perhaps(source)
+            end
+
+            if cwm.Eos.is_set(m.flags) then
+              // Process EOS
+              @printf[I32](("ConnectorSource[%s] received EOS for stream_id"
+                + " %s\n").cstring(), _source_name.string().cstring(),
+                m.stream_id.string().cstring())
+              // 1. remove state from _active_streams
+              try _active_streams.remove(m.stream_id)? end
+              // 2. add state to _pending_close
+              ifdef "resilience" then
+                _pending_close.update(m.stream_id, s)
+              else
+                // respond immediately
+                _send_reply(source, cwm.AckMsg(0, [(s.id, s.last_seen)]))
+                _listener.streams_relinquish([StreamTuple(s.id, s.name,
+                  s.last_seen)])
+                // TODO [source-migration] remove stream from active
+              end
+              return true
+            elseif cwm.Boundary.is_set(m.flags) then
+              // TODO [post-source-migration] what's supposed to happen here?
+              return true
+            else // not an EOS and not a boundary
+              // process message
+              try
+                // get bytes content of message
+                let bytes = match (m.message as cwm.MessageBytes)
+                | let str: String      => str.array()
+                | let b: Array[U8] val => b
+                end
+
+                // decode bytes using handler, if bytes not None
+                let decoded = if bytes.size() == 0 then
```

Comment:
Is it valid for bytes.size() to be 0?  Can we optimistically call decode in a try block and then do something if there's an error (in which case we can still check if bytes.size() is 0 if that differentiates a particular problem case).